File: app/src/main/java/com/jgm90/cloudmusic/fragments/SearchFragment.java
package com.jgm90.cloudmusic.fragments;

import android.os.Bundle;
import android.os.Parcelable;
import androidx.core.view.MenuItemCompat;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.SearchView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.jgm90.cloudmusic.R;
import com.jgm90.cloudmusic.R2;
import com.jgm90.cloudmusic.adapters.SearchAdapter;
import com.jgm90.cloudmusic.interfaces.RestInterface;
import com.jgm90.cloudmusic.models.SongModel;
import com.jgm90.cloudmusic.utils.Divider;
import com.jgm90.cloudmusic.utils.NetworkHelper;
import com.jgm90.cloudmusic.utils.RestClient;
import com.jgm90.cloudmusic.utils.SharedUtils;
import com.jgm90.cloudmusic.widgets.VulgryMessageView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchFragment extends BaseFragment implements SearchView.OnQueryTextListener {

    public final static String LIST_STATE_KEY = "recycler_list_state";
    public final static String LIST_ARRAY = "recycler_list_model";

    @BindView(R2.id.message_view)
    public VulgryMessageView message_view;
    @BindView(R2.id.loader)
    public MaterialProgressBar loader;
    @BindView(R2.id.rv_search)
    public RecyclerView mRecyclerView;
    public String search_query;
    Parcelable listState;
    private SearchAdapter mAdapter;
    private List<SongModel> mModel;
    private AppCompatActivity activity;
    private RecyclerView.LayoutManager mLayoutManager;
    // unbinder
    private Unbinder unbinder;
    private SearchView searchView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_search, container, false);
        activity = (AppCompatActivity) getActivity();
        unbinder = ButterKnife.bind(this, rootView);
        mLayoutManager = new LinearLayoutManager(activity);
        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.addItemDecoration(new Divider(activity));
        mRecyclerView.getItemAnimator().setAddDuration(SharedUtils.rv_anim_duration);
        mRecyclerView.setAdapter(null);
        mModel = new ArrayList<>();
        setHasOptionsMenu(true);
        if (savedInstanceState != null) {
            listState = savedInstanceState.getParcelable(LIST_STATE_KEY);
            mModel = savedInstanceState.getParcelableArrayList(LIST_ARRAY);
            mAdapter = new SearchAdapter(mModel, activity);
            mRecyclerView.setAdapter(mAdapter);
            mAdapter.notifyDataSetChanged();
        }
        if (mModel.size() == 0) {
            SharedUtils.showMessage(message_view, R.drawable.ic_search_black_24dp, R.string.search_message);
        }
        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        listState = mLayoutManager.onSaveInstanceState();
        outState.putParcelable(LIST_STATE_KEY, listState);
        outState.putParcelableArrayList(LIST_ARRAY, new ArrayList<>(mModel));
    }

    @Override
    public void onResume() {
        super.onResume();
        if (listState != null) {
            mLayoutManager.onRestoreInstanceState(listState);
        }
    }

    @OnClick(R2.id.message_view)
    public void reload() {
        if (search_query.length() > 0) {
            if (NetworkHelper.isOnline(getContext())) {
                message_view.setVisibility(View.GONE);
                mModel.clear();
                getResult(search_query);
                searchView.clearFocus();
            } else {
                SharedUtils.showMessage(message_view, R.drawable.ic_cloud_off_black_24dp, R.string.error_offline);
            }
        }
    }

    public void getResult(String query) {
        loader.setVisibility(View.VISIBLE);
        try {
            RestInterface api = RestClient.build(SharedUtils.server);
            Call<List<SongModel>> call = api.getSongs(query, 1, 50);
            call.enqueue(new Callback<List<SongModel>>() {
                @Override
                public void onResponse(Call<List<SongModel>> call, Response<List<SongModel>> response) {
                    loader.setVisibility(View.GONE);
                    if (response.isSuccessful()) {
                        if (response.body() != null && response.body().size() > 0) {
                            for (int i = 0; i < response.body().size(); i++) {
                                mModel = response.body();
                            }
                            mAdapter = new SearchAdapter(mModel, activity);
                            mRecyclerView.setAdapter(mAdapter);
                            mAdapter.notifyItemChanged(0);
                        }
                    } else {
                        SharedUtils.showMessage(message_view, R.drawable.ic_error_black_24dp, R.string.error_retrofit);
                    }
                }

                @Override
                public void onFailure(Call<List<SongModel>> call, Throwable t) {
                    loader.setVisibility(View.GONE);
                    SharedUtils.showMessage(message_view, R.drawable.ic_cloud_off_black_24dp, R.string.error_server);
                }
            });
        } catch (Exception e) {
            loader.setVisibility(View.GONE);
            e.printStackTrace();
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_search, menu);
        final MenuItem item = menu.findItem(R.id.action_search);
        searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setQueryHint("Buscar");
        searchView.setOnQueryTextListener(this);
        if (!TextUtils.isEmpty(search_query)) {
            searchView.setQuery(search_query, false);
            searchView.setIconified(false);
            searchView.clearFocus();
        }
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        search_query = newText;
        return false;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        mModel.clear();
        mRecyclerView.setAdapter(null);
        reload();
        return true;
    }
}

File: app/src/main/java/com/jgm90/cloudmusic/feature/search/presentation/SearchFragment.kt
package com.jgm90.cloudmusic.feature.search.presentation

import android.app.DownloadManager
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.HorizontalDivider
import androidx.compose.runtime.getValue
import androidx.core.view.MenuItemCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.jgm90.cloudmusic.R
import com.jgm90.cloudmusic.activities.NowPlayingActivity
import com.jgm90.cloudmusic.databinding.FragmentSearchBinding
import com.jgm90.cloudmusic.dialogs.AddToPlaylistDialog
import com.jgm90.cloudmusic.events.DownloadEvent
import com.jgm90.cloudmusic.feature.search.presentation.viewmodel.SearchViewModel
import com.jgm90.cloudmusic.ui.theme.CloudMusicTheme
import io.nubit.cloudmusic.designsystem.component.EmptyState
import io.nubit.cloudmusic.designsystem.component.Loader
import io.nubit.cloudmusic.designsystem.component.SongItem
import org.greenrobot.eventbus.EventBus

class SearchFragment : Fragment(), SearchView.OnQueryTextListener {
    private var _binding: FragmentSearchBinding? = null
    private val binding get() = _binding!!
    private val viewModel by viewModels<SearchViewModel>()
    private var searchQuery: String? = null
    private var searchView: SearchView? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentSearchBinding.inflate(inflater, container, false)
        setHasOptionsMenu(true)
        val rootView = binding.root
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.composeView.setContent {
            CloudMusicTheme {
                val state by viewModel.uiState.collectAsStateWithLifecycle()
                Loader(state.isLoading) {
                    if (state.error != null) {
                        EmptyState(
                            textRes = R.string.error_retrofit,
                            imageRes = R.drawable.ic_error_black_24dp,
                        )
                    } else if (state.searchResults.isEmpty()) {
                        EmptyState(
                            textRes = R.string.search_message,
                            imageRes = R.drawable.ic_search_black_24dp,
                        )
                    }

                    LazyColumn {
                        itemsIndexed(state.searchResults) { index, song ->
                            SongItem(
                                imageUrl = song.getCoverThumbnail(),
                                songName = song.name,
                                artistName = song.artist,
                                albumName = song.album,
                                onClick = {
                                    val intent = Intent(
                                        context,
                                        NowPlayingActivity::class.java
                                    )
                                    intent.putExtra("SONG_INDEX", index)
                                    NowPlayingActivity.audioList = state.searchResults
                                    requireContext().startActivity(intent)
                                },
                                onDownloadClick = {
                                    EventBus.getDefault().postSticky(
                                        DownloadEvent(
                                            true,
                                            DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED,
                                            song.getAudioUrl(),
                                            song.name,
                                            song.getFileName()
                                        )
                                    )
                                },
                                onAddToPlaylistClick = {
                                    AddToPlaylistDialog(context).show(song)
                                }
                            )
                            HorizontalDivider()
                        }
                    }
                }
            }
        }
    }

    private fun reload() {
        searchQuery?.let {
            viewModel.search(it)
            searchView!!.clearFocus()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_search, menu)
        val item = menu.findItem(R.id.action_search)
        searchView = MenuItemCompat.getActionView(item) as SearchView
        searchView!!.queryHint = "Buscar"
        searchView!!.setOnQueryTextListener(this)
        if (!TextUtils.isEmpty(searchQuery)) {
            searchView!!.setQuery(searchQuery, false)
            searchView!!.isIconified = false
            searchView!!.clearFocus()
        }
    }

    override fun onQueryTextChange(newText: String): Boolean {
        searchQuery = newText
        return false
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        reload()
        return true
    }
}

